front/src/modules/trainschedule/components/ManageTrainSchedule/helpers/powerRestrictionSelector.ts

```diff
@@ -0,0 +1,91 @@
+
+import NO_POWER_RESTRICTION from '../consts';
+import type { PowerRestrictionWarning } from '../types';
+
+/**
+ * Return the power restriction codes of the rolling stock by mode
+ *
+ * ex: { "1500V": ["C1US", "C2US"], "25000V": ["M1US"], "thermal": []}
+ */
+const getRollingStockPowerRestrictionsByMode = (
+  rollingStockModes: RollingStock['effort_curves']['modes']
+): { [mode: string]: string[] } => {
+  const curvesModesKey = Object.keys(rollingStockModes);
+
+  return reduce(
+    curvesModesKey,
+    (result, mode) => {
+      const powerCodes = rollingStockModes[mode].curves.map(
+        (curve) => curve.cond.power_restriction_code
+      );
+      compact(uniq(powerCodes));
+      return {
+        ...result,
+        [mode]: powerCodes,
+      };
+    },
+    {}
+  );
+};
+
+/**
+ * Depending on the electrification on a train's path and the power restriction codes selected by the user,
+ * returns the possible warnings based on the incompatibilty between them on each interval
+ * @returns A dictionnary of warnings ordered by power restriction codes
+ */
+export const getPowerRestrictionsWarnings = (
+  powerRestrictionRanges: RangedValue[],
+  pathElectrificationRanges: RangedValue[],
+  rollingStockModes: RollingStock['effort_curves']['modes']
+): Dictionary<PowerRestrictionWarning[]> => {
+  const powerRestrictionsByMode = getRollingStockPowerRestrictionsByMode(rollingStockModes);
+  const warnings: PowerRestrictionWarning[] = [];
+
+  powerRestrictionRanges.forEach((powerRestrictionRange) => {
+    // find path ranges crossed or included in the power restriction range
+    pathElectrificationRanges.forEach((pathElectrificationRange) => {
+      // no intersection between the path range and the power restriction range
+      if (
+        pathElectrificationRange.end <= powerRestrictionRange.begin ||
+        powerRestrictionRange.end <= pathElectrificationRange.begin
+      )
+        return;
+      // check restriction is compatible with the path range's electrification mode
+      const isInvalid = !powerRestrictionsByMode[pathElectrificationRange.value].includes(
+        powerRestrictionRange.value
+      );
+      if (isInvalid) {
+        const invalidZoneBegin = Math.round(
+          pathElectrificationRange.begin < powerRestrictionRange.begin
+            ? powerRestrictionRange.begin
+            : pathElectrificationRange.begin
+        );
+        const invalidZoneEnd = Math.round(
+          powerRestrictionRange.end < pathElectrificationRange.end
+            ? powerRestrictionRange.end
+            : pathElectrificationRange.end
+        );
+
+        warnings.push({
+          powerRestrictionCode:
+            powerRestrictionRange.value !== NO_POWER_RESTRICTION
+              ? powerRestrictionRange.value
+              : NO_POWER_RESTRICTION,
+          electrification:
+            powerRestrictionRange.value !== NO_POWER_RESTRICTION
+              ? pathElectrificationRange.value
+              : undefined,
```

Comment:
```suggestion
          ...(powerRestrictionRange.value !== NO_POWER_RESTRICTION
            ? {
                powerRestrictionCode: powerRestrictionRange.value,
                electrification: pathElectrificationRange.value,
              }
            : { powerRestrictionCode: NO_POWER_RESTRICTION }),
```

Reply:
Done

Reply:
I kept it anyway even without the type changing